pkg/codegen/hcl2/rewrite_convert.go

```diff
@@ -116,5 +118,105 @@ func RewriteConversions(x model.Expression, to model.Type) model.Expression {
 	if to.AssignableFrom(x.Type()) && sameSchemaTypes(to, x.Type()) {
 		return x
 	}
+	// If we can convert a primitive value in place, do so.
+	if value, ok := convertPrimitiveValues(x, to); ok {
+		return value
+	}
 	return NewConvertCall(x, to)
 }
+
+// convertPrimitiveValues returns a new expression if the given expression can be converted to another primitive type
+// (bool, int, number, string) that matches the target type.
+func convertPrimitiveValues(from model.Expression, to model.Type) (model.Expression, bool) {
+	value := cty.NilVal
+	switch resolveUnderlyingType(to) {
+	case model.BoolType:
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if b, err := strconv.ParseBool(stringLiteral); err == nil {
+				value = cty.BoolVal(b)
+			}
+		}
+	case model.IntType:
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if i, err := strconv.ParseInt(stringLiteral, 10, 64); err == nil {
+				value = cty.NumberIntVal(i)
+			}
+		}
+	case model.NumberType:
+		if stringLiteral, ok := extractStringValue(from); ok {
+			if f, err := strconv.ParseFloat(stringLiteral, 64); err == nil {
+				value = cty.NumberFloatVal(f)
+			}
+		}
+	case model.StringType:
+		stringValue := convertLiteralToString(from)
+		if stringValue != "" {
+			value = cty.StringVal(stringValue)
+		}
+	}
+	if value == cty.NilVal {
+		return nil, false
+	}
+
+	stringLiteral := model.LiteralValueExpression{Value: value}
```

Comment:
String literals in expressions are a bit weird: in most contexts, they are represented as single-part literals. This includes any position in which we will insert a conversion, so this should be:

```suggestion
	stringLiteral := &model.TemplateExpression{
		Parts: []model.Expression{&model.LiteralValueExpression{
			Value: value,
		}},
	}
	stringLiteral.SetLeadingTrivia(from.GetLeadingTrivia())
	stringLiteral.SetTrailingTrivia(from.GetTrailingTrivia())
	return stringLiteral, true
```